### ISSUE TITLE
upgrade possibilty? and thx

### ISSUE THREAD

**Issue by komodensys** [rtestardi/usbte#6]
hello friend thx for hard work and lovely scope..   is it possible to make more analog channels in the same page (either with other boards or even spliting the 5 adcs  into 5 channels? that would be great??

**Comment by rtestardi** [comment]
You are very welcome!  You can open two web-pages and switch them to "background" mode with the button so that they can both run and capture,, even if only one is visible at a time (ordinarily javascript pauses the non-foreground web-page, which totally breaks the trigger logic) -- I realize it is a bit clunky to do that.

I think a couple of analog channels from two different flea-scopes on the same page is a great possible enhancement...  Possibly in place of digital capture...  There are some challenges as well -- with gathering twice as much data from USB, and handling independent triggers, and handling two different sets of calibration values, but it could be doable.

As for 5 adc's into 5 channels, that would require very different hardware.  And unfortunately the "raw adc input" is quite low impedance, (like 10k ohm or less) making it probably unsuitable for any kind of measurements -- and adding 5x op-amps as front-ends for 5 adc's starts to get pricey...

Possibly for the future mosquito-scope if I can do it on dsPIC33A, which should have 2x 40 Msps adc's, it might make sense to have 2 channels of 40 Msps -- that's be a nice balance of cost and performance.

**Comment by komodensys** [comment]
by the looks the bea-scope will be pure honey....lol  
Thx for answer and I understand the limitations, almost all diy scopes are 1 or 2 channels thats why I was looking for 3 or 4 max channels
For automotive work theres no need for high sampling but more channels are appreciated
when we buy the fabricated pcb its minimum 4 pcbs so it would be great improvement more channels 
love the simplistic way of altering the gui page (did it in notepad, decreasing the font sizes for better split screenl) and the easy of taking measurements but that all change with more channels I know
Anyway you already have a fan just by making it opensource and being a teacher that teaches and cares about the students 
You deserve to teach in the university as your dream...   Big thanks from Portugal

**Comment by rtestardi** [comment]
I also love the automotive target -- I have done some you can see [here](https://www.zx6r.com/threads/running-with-the-tank-off-ignition-injector-adventure.120654/?post_id=1915926#post-1915926) -- I totally agree more channels (rather than more sample rate) gets nicer.  I do hope to make it to teaching somehow -- it is especially nice when folks are grateful, so thank you for that!

**Comment by UltraBlackLinux** [comment]
As I said on hacker news, I think it makes sense to keep any flea-scope successors at the same form-factor, with just a single analog input since you can very easily just hook up several scopes to get more channels, which is further justified by the low cost of the device. What I'm hoping to see with a potential successor is a faster MCU (for even higher-frequency signals) and a way to chain scopes together and have them all output data over a single master scope's usb output.

To make hooking up more scopes easier I've actually started development of a more modular interface with the goal of being able to simultaneously manage pretty much any amount of scopes in one interface while also making the code much easier to understand and maintain overall. ;)

Anyway, you can track my progress here: https://chonkyrabbit.eu/git/flea-scope-ui.git/tree/?h=main

Thanks for this lovely little device!

(I'm currently still trying to figure out if reading from multiple serial devices in the same tab is even possible - the info on this seems to be rather scarse)

**Comment by rtestardi** [comment]
Nice -- the challenge is the calibration model -- unfirtunately, the way I wrote it is is somewhat of a dance between the gui and the firmware -- the gui calculates calibration values and stores them in the flea-scope flash, and then queries the values from flash to adjust the traces on display...  I should have done it all in firmware so the traces came out of the flea-scope already calibrated, but there was never a good time to do that (because it would break all older scopes if the gui changed).  That is by far the biggest flaw in the design (that would be fixed in a new version, like mosquito-scope).  Possibly calibration could be done both ways, and a scope that already self-calibrated its output (by snooping writes to the flash calibration values) could report 0 calibration values to the gui, so the gui did not try and do calibration a second time?

**Comment by UltraBlackLinux** [comment]
I don't think this is necessarily a bad decision. As you've also noted yourself the scope requires somewhat frequent recalibration because of changes in temperature. Writing new data to the flash every few hours would no doubt wear the flash down and likely ultimately kill it (although I don't know if the maximum write count really is this low, but I hope you get my point). What I was just thinking of is that the scope could potentially actually do the calibration itself, by using maybe a transistor/FET of some kind to could connect 3.3v and then ground to the scope's own analog input and then do the calibration internally based off that. That would get rid of the requirements for flash-dependent calibrations entirely.

Also another thought I had earlier today is that scope chaining could actually already be done using the current flea-scope hardware since many of the analog basic-reserved inputs are unused in scope mode. I was thinking that the scopes could maybe just pipe their own serial outputs alongside any that comes in via the pins out from their own analog pins. This could also allow for trigger sends without a dedicated pin. The downside is that you'd need a new way of communicating over serial that identifies all scopes, maybe with some sort of header at the top of each data block.

But regarding the problem of changing the way the scope calibrates this rewrite could actually prove quite helpful here as well since the at least somewhat modular way I implemented most features should (hopefully) allow me to look at e.g. the usb PID and change the parse behavior based on that.
